### DISCUSSION TITLE
Discarding detections?

### DISCUSSION THREAD

**Discussion by jkrauska** [tphakala/birdnet-go#332] [General]
I'm listening live and definately hear these birds. 

Why are p89s being discarded?
```
2024/12/19 13:32:42 New detection: Chestnut-backed Chickadee with confidence: 0.88, source: malgo
2024/12/19 13:32:52 Discarding detection of chestnut-backed chickadee from source malgo as false positive, matched 1/10 times

2024/12/19 13:35:37 New detection: House Finch with confidence: 0.89, source: malgo
2024/12/19 13:35:48 Discarding detection of house finch from source malgo as false positive, matched 1/10 times
```

**Comment by jkrauska** [comment]
More details.  Is the match count not tracked in a DB?

```
2024/12/19 13:39:25 New detection: Chestnut-backed Chickadee with confidence: 0.81, source: malgo
2024/12/19 13:39:25 Updating detection: Chestnut-backed Chickadee with confidence: 0.83, source: malgo, count: 2
2024/12/19 13:39:26 Updating detection: Chestnut-backed Chickadee with confidence: 0.87, source: malgo, count: 3
2024/12/19 13:39:35 Discarding detection of chestnut-backed chickadee from source malgo as false positive, matched 7/10 times

2024/12/19 13:39:56 New detection: Chestnut-backed Chickadee with confidence: 0.63, source: malgo
2024/12/19 13:39:56 Updating detection: Chestnut-backed Chickadee with confidence: 0.92, source: malgo, count: 2
2024/12/19 13:40:06 Discarding detection of chestnut-backed chickadee from source malgo as false positive, matched 7/10 times
2024/12/19 13:40:40 New detection: Chestnut-backed Chickadee with confidence: 0.50, source: malgo
2024/12/19 13:40:51 Discarding detection of chestnut-backed chickadee from source malgo as false positive, matched 1/10 times


2024/12/19 13:42:03 New detection: Chestnut-backed Chickadee with confidence: 0.64, source: malgo
2024/12/19 13:42:04 Updating detection: Chestnut-backed Chickadee with confidence: 0.67, source: malgo, count: 2
2024/12/19 13:42:05 Updating detection: Chestnut-backed Chickadee with confidence: 0.71, source: malgo, count: 3
2024/12/19 13:42:14 Discarding detection of chestnut-backed chickadee from source malgo as false positive, matched 3/10 times
2024/12/19 13:42:38 New detection: Chestnut-backed Chickadee with confidence: 0.56, source: malgo
2024/12/19 13:42:48 Discarding detection of chestnut-backed chickadee from source malgo as false positive, matched 2/10 times
2024/12/19 13:42:54 Human detected with confidence 0.056/0.050 from source malgo
2024/12/19 13:43:04 New detection: Orange-crowned Warbler with confidence: 0.82, source: malgo
2024/12/19 13:43:07 New detection: Chestnut-backed Chickadee with confidence: 0.45, source: malgo
2024/12/19 13:43:08 Updating detection: Chestnut-backed Chickadee with confidence: 0.51, source: malgo, count: 2
2024/12/19 13:43:08 Updating detection: Chestnut-backed Chickadee with confidence: 0.63, source: malgo, count: 3
2024/12/19 13:43:15 Discarding detection of orange-crowned warbler from source malgo as false positive, matched 1/10 times
2024/12/19 13:43:18 Discarding detection of chestnut-backed chickadee from source malgo as false positive, matched 3/10 times
```

**Reply by tphakala** [reply]
This behavior is controlled by the "deep detection" feature, which tries to eliminate false positives by requiring multiple results (hits) of species by the BirdNET AI model from a 3-second audio segment before accepting detection.

In the logs you provided, the minimum detection count is 10, and the chestnut-backed chickadee was detected only 3 times with confidence exceeding the configured confidence threshold.

The detection count threshold is controlled by the BirdNET overlap value. The code to calculate the minDetections value is in analysis/processor/processor.go:

```go
// pendingDetectionsFlusher runs a goroutine that periodically checks the pending detections
// and flushes them to the worker queue if their deadline has passed.
func (p *Processor) pendingDetectionsFlusher() {
	// Determine minDetections based on Settings.BirdNET.Overlap
	var minDetections int

	// Calculate segment length based on overlap setting, minimum 0.1 seconds
	segmentLength := math.Max(0.1, 3.0-p.Settings.BirdNET.Overlap)
	// Calculate minimum detections needed based on segment length, at least 1
	minDetections = int(math.Max(1, 3/segmentLength))
```

I have been thinking of adding configuration options that allow more control over the minDetections value, but then again, I also don't want to make things too complicated for users. However, setting BirdNET overlap value to 0 accepts detections on first hit.

**Reply by jkrauska** [reply]
@tphakala is the "minimum detection count" a global count?  I'm guessing that upon daemon restart that these get zero'd out and need to start over?

if so, can we store in sqllite to avoid a slow start?

**Reply by Fotguedes** [reply]
Another alternative would be to add an option to save false positives for later review, instead of discarding them immediately.

Additionally, it might be useful to introduce a threshold range for manual verification. For example:
- If the overlap parameter is set to detect 10 hits, but a species was detected in more than 5 hits (yet still below the minimum threshold), it could be flagged for manual review instead of being automatically discarded.

In my case, I had overlap set to 2, and I noticed that it often excluded 1/2 hits even when I was certain—based on listening to the stream—that the species was actually present.

This could also be beneficial for future improvements in model training, as it would allow for better validation based on user-reviewed identifications (both correct and incorrect). Having a reviewed dataset could help refine the detection accuracy and reduce misclassifications over time.

**Reply by tphakala** [reply]
good point, need to consider this
